File: app/src/lib/api/queries/time-tracking.ts
import { queryOptions } from "@tanstack/react-query";
import { api } from "../api";
import dayjs from "dayjs";

export type DateRangeQuery = {
  from: string;
  to: string;
};

export type TimeEntriesQuery = DateRangeQuery & {
  unique?: boolean;
};

const timeTrackingQueryKeys = {
  projectsBase: ["time-tracking", "projects"] as const,
  activitiesBase: ["time-tracking", "activities"] as const,
  timerBase: ["time-tracking", "timer"] as const,
  timeEntriesBase: ["time-tracking", "time-entries"] as const,
  absenceEntriesBase: ["time-tracking", "absences"] as const,
  absenceTypesBase: ["time-tracking", "absence-types"] as const,
  absenceChildrenBase: ["time-tracking", "absence-children"] as const,
  absenceDayDefaultsBase: ["time-tracking", "absence-day-defaults"] as const,
  timeInfoBase: ["time-tracking", "time-info"] as const,
  timeEntryDayStatusesBase: [
    "time-tracking",
    "time-entry-day-statuses",
  ] as const,
  activities: (projectId: string) =>
    [...timeTrackingQueryKeys.activitiesBase, projectId] as const,
  timer: () => [...timeTrackingQueryKeys.timerBase, "get"] as const,
  timerHistory: () => [...timeTrackingQueryKeys.timerBase, "history"] as const,
  timeEntries: (query?: TimeEntriesQuery) =>
    [
      ...timeTrackingQueryKeys.timeEntriesBase,
      query?.from,
      query?.to,
      query?.unique,
    ] as const,
  absenceEntries: (query: DateRangeQuery) =>
    [
      ...timeTrackingQueryKeys.absenceEntriesBase,
      query.from,
      query.to,
    ] as const,
  absenceDayDefaults: (query: DateRangeQuery) =>
    [
      ...timeTrackingQueryKeys.absenceDayDefaultsBase,
      query.from,
      query.to,
    ] as const,
  timeInfo: (query?: DateRangeQuery) =>
    [...timeTrackingQueryKeys.timeInfoBase, query?.from, query?.to] as const,
  timeEntryDayStatuses: (query: DateRangeQuery) =>
    [
      ...timeTrackingQueryKeys.timeEntryDayStatusesBase,
      query.from,
      query.to,
    ] as const,
};

export function parseTimeEntriesQueryKey(
  queryKey: readonly unknown[],
): TimeEntriesQuery | null {
  const [scope, resource, from, to, unique] = queryKey;
  if (
    scope !== "time-tracking" ||
    resource !== "time-entries" ||
    typeof from !== "string" ||
    typeof to !== "string"
  ) {
    return null;
  }

  return {
    from,
    to,
    unique: typeof unique === "boolean" ? unique : undefined,
  };
}

export function parseAbsenceEntriesQueryKey(
  queryKey: readonly unknown[],
): DateRangeQuery | null {
  const [scope, resource, from, to] = queryKey;
  if (
    scope !== "time-tracking" ||
    resource !== "absences" ||
    typeof from !== "string" ||
    typeof to !== "string"
  ) {
    return null;
  }

  return { from, to };
}

export function parseTimeInfoQueryKey(
  queryKey: readonly unknown[],
): DateRangeQuery | null {
  const [scope, resource, from, to] = queryKey;
  if (
    scope !== "time-tracking" ||
    resource !== "time-info" ||
    typeof from !== "string" ||
    typeof to !== "string"
  ) {
    return null;
  }

  return { from, to };
}

function sortByName<T>(items: Array<T>, getName: (item: T) => string) {
  return items.toSorted((a, b) =>
    getName(a).localeCompare(getName(b), undefined, {
      numeric: true,
      sensitivity: "base",
    }),
  );
}

export const timeTrackingQueries = {
  projectsBaseKey: timeTrackingQueryKeys.projectsBase,
  activitiesBaseKey: timeTrackingQueryKeys.activitiesBase,
  timeEntriesBaseKey: timeTrackingQueryKeys.timeEntriesBase,
  absenceEntriesBaseKey: timeTrackingQueryKeys.absenceEntriesBase,
  absenceTypesBaseKey: timeTrackingQueryKeys.absenceTypesBase,
  absenceChildrenBaseKey: timeTrackingQueryKeys.absenceChildrenBase,
  absenceDayDefaultsBaseKey: timeTrackingQueryKeys.absenceDayDefaultsBase,
  timeInfoBaseKey: timeTrackingQueryKeys.timeInfoBase,
  listProjects: () =>
    queryOptions({
      queryKey: timeTrackingQueries.projectsBaseKey,
      queryFn: async () =>
        sortByName(
          await api.get("time-tracking/projects").json<Array<Project>>(),
          (project) => project.projectName,
        ),
      staleTime: 60 * 60 * 1000,
      gcTime: 24 * 60 * 60 * 1000,
    }),
  connectionStatus: () =>
    queryOptions({
      queryKey: ["time-tracking", "connection"],
      queryFn: async () =>
        api
          .get("time-tracking/connection")
          .json<TimeTrackingConnectionStatus>(),
    }),
  listActivities: (projectId: string) =>
    queryOptions({
      queryKey: timeTrackingQueryKeys.activities(projectId),
      queryFn: async () =>
        sortByName(
          await api
            .get(`time-tracking/projects/${projectId}/activities`)
            .json<Array<Activity>>(),
          (activity) => activity.activityName,
        ),
      staleTime: 60 * 60 * 1000,
      gcTime: 24 * 60 * 60 * 1000,
    }),
  timerBaseKey: timeTrackingQueryKeys.timerBase,
  getTimer: () =>
    queryOptions({
      queryKey: timeTrackingQueryKeys.timer(),
      queryFn: async () =>
        api.get("time-tracking/timer").json<GetTimerResponse>(),
    }),
  timerHistory: () =>
    queryOptions({
      queryKey: timeTrackingQueryKeys.timerHistory(),
      queryFn: async () =>
        api.get("time-tracking/timer-history").json<Array<TimerHistoryEntry>>(),
    }),
  timeInfo: (query?: DateRangeQuery) =>
    queryOptions({
      queryKey: timeTrackingQueryKeys.timeInfo(query),
      queryFn: async () => {
        return api
          .get("time-tracking/time-info", {
            searchParams: query ?? {
              from: dayjs().startOf("month").format("YYYY-MM-DD"),
              to: dayjs().endOf("month").format("YYYY-MM-DD"),
            },
          })
          .json<WeeklyStats>();
      },
    }),
  timeEntries: (query?: TimeEntriesQuery) =>
    queryOptions({
      queryKey: timeTrackingQueryKeys.timeEntries(query),
      queryFn: async () => {
        return api
          .get("time-tracking/time-entries", {
            searchParams: query ?? {
              from: dayjs().startOf("month").format("YYYY-MM-DD"),
              to: dayjs().endOf("month").format("YYYY-MM-DD"),
            },
          })
          .json<Array<TimeEntry>>();
      },
      staleTime: query?.unique ? 5 * 60 * 1000 : undefined,
      gcTime: query?.unique ? 30 * 60 * 1000 : undefined,
    }),
  absenceEntries: (query: DateRangeQuery) =>
    queryOptions({
      queryKey: timeTrackingQueryKeys.absenceEntries(query),
      queryFn: async () =>
        api
          .get("time-tracking/absences", {
            searchParams: query,
          })
          .json<Array<AbsenceEntry>>(),
    }),
  absenceTypes: () =>
    queryOptions({
      queryKey: timeTrackingQueries.absenceTypesBaseKey,
      queryFn: async () =>
        api.get("time-tracking/absence-types").json<Array<AbsenceTypeOption>>(),
      staleTime: 60 * 60 * 1000,
      gcTime: 24 * 60 * 60 * 1000,
    }),
  absenceChildren: () =>
    queryOptions({
      queryKey: timeTrackingQueries.absenceChildrenBaseKey,
      queryFn: async () =>
        api
          .get("time-tracking/absence-children")
          .json<Array<AbsenceChild>>(),
      staleTime: 5 * 60 * 1000,
      gcTime: 60 * 60 * 1000,
    }),
  absenceDayDefaults: (query: DateRangeQuery) =>
    queryOptions({
      queryKey: timeTrackingQueryKeys.absenceDayDefaults(query),
      queryFn: async () =>
        api
          .get("time-tracking/absence-day-defaults", {
            searchParams: query,
          })
          .json<Array<AbsenceDayDefault>>(),
    }),
  timeEntryDayStatusesBaseKey: timeTrackingQueryKeys.timeEntryDayStatusesBase,
  timeEntryDayStatuses: (query: DateRangeQuery) =>
    queryOptions({
      queryKey: timeTrackingQueryKeys.timeEntryDayStatuses(query),
      queryFn: async () => {
        return api
          .get("time-tracking/time-entry-day-statuses", {
            searchParams: query,
          })
          .json<Array<TimeEntryDayStatus>>();
      },
      staleTime: 2 * 60 * 1000,
      gcTime: 30 * 60 * 1000,
    }),
  adminMappings: () =>
    queryOptions({
      queryKey: ["time-tracking", "admin", "kleer-users"],
      queryFn: async () =>
        api
          .get("time-tracking/admin/kleer-users")
          .json<TimeTrackingAdminMappings>(),
    }),
};

export type GetTimerResponse = {
  timer: TimerResponse | null;
};

export type SaveTimerResponse = {
  entry: TimeEntry;
  timer: TimerResponse | null;
};

export type TimeTrackingConnectionStatus = {
  connected: boolean;
  providerUserId: string | null;
  providerUserEmail: string | null;
  providerUserName: string | null;
};

/** Active timer response. */
export type TimerResponse = {
  startTime: string;
  projectId: string | null;
  projectName: string | null;
  activityId: string | null;
  activityName: string | null;
  note: string;
  hours: number;
  minutes: number;
  seconds: number;
};

/** Timer history entry from the database. */
export type TimerHistoryEntry = {
  id: number;
  registrationId: string | null;
  userId: number;
  startTime: string;
  endTime: string | null;
  projectId: string | null;
  projectName: string | null;
  activityId: string | null;
  activityName: string | null;
  note: string | null;
  createdAt: string;
};

export type WeeklyStats = {
  workedHours: number;
  scheduledHours: number;
  remainingHours: number;
  absenceHours: number;
  coveredHours: number;
  periodFlexHours: number;
};

export type Project = {
  projectId: string;
  projectName: string;
};

export type Activity = {
  activity: string;
  activityName: string;
};

export type TimeEntry = {
  registrationId: string;
  projectId: string;
  projectName: string;
  activityId: string;
  activityName: string; // activityId is the raw activity code used when (re)starting timers
  date: string;
  hours: number;
  note: string | null;
  startTime: string | null;
  endTime: string | null;
  weekNumber: number;
  status: TimeEntryStatus;
};

export type TimeEntryStatus = "open" | "approved" | "certified";

export type TimeEntryDayStatus = {
  date: string;
  status: TimeEntryStatus;
};

export type AbsenceType =
  | "sick"
  | "vacation"
  | "leaveOfAbsence"
  | "leaveOfAbsenceVacationEarned"
  | "parentalLeave"
  | "childcare"
  | "closeRelativeCare"
  | "paternityLeave"
  | "furlough"
  | "otherLeave"
  | "otherLeaveVacationNotEarned";

export type ManagedAbsenceType = AbsenceType;

export const absenceTypeLabels = {
  sick: "Sick (Sjuk)",
  vacation: "Vacation (Semester)",
  leaveOfAbsence: "Leave of absence (Tjänstledig)",
  leaveOfAbsenceVacationEarned:
    "Leave of absence, vacation earned (Tjänstledig (Semestergrundande))",
  parentalLeave: "Parental leave (Föräldraledighet)",
  childcare: "Childcare (VAB)",
  closeRelativeCare: "Close relative care (Vård av nära anhörig)",
  paternityLeave: "Paternity leave (10 dagar vid barns födelse)",
  furlough: "Furlough (Permission)",
  otherLeave: "Other leave (Övrig frånvaro)",
  otherLeaveVacationNotEarned:
    "Other leave, vacation not earned (Övrig frånvaro (Semestergrundande))",
} satisfies Record<AbsenceType, string>;

export type AbsenceTypeOption = {
  absenceType: ManagedAbsenceType;
  absenceTypeLabel: string;
};

export type AbsenceChild = {
  name: string;
  birthDate: string | null;
};

export type AbsenceEntry = {
  absenceId: string;
  date: string;
  hours: number;
  absenceType: AbsenceType;
  absenceTypeLabel: string;
  child: string | null;
  comment: string | null;
  managed: boolean;
  deletable: boolean;
};

export type AbsenceDayDefault = {
  date: string;
  scheduledHours: number;
};

export type TimeTrackingAdminMappings = {
  users: Array<TimeTrackingAdminUser>;
  kleerUsers: Array<TimeTrackingAdminKleerUser>;
  links: Array<TimeTrackingAdminUserLink>;
};

export type TimeTrackingAdminUser = {
  id: number;
  email: string;
  fullName: string;
};

export type TimeTrackingAdminKleerUser = {
  providerUserId: string;
  foreignId: string | null;
  internalId: string | null;
  name: string;
  email: string | null;
  active: boolean;
  mappedUserId: number | null;
  mappedUserEmail: string | null;
  mappedUserName: string | null;
  lastSyncedAt: string;
};

export type TimeTrackingAdminUserLink = {
  id: number;
  userId: number;
  providerUserId: string;
  providerUserEmail: string | null;
  providerUserName: string | null;
  updatedAt: string;
};

File: toki-api/src/adapters/inbound/http/responses/time_tracking.rs
use serde::Serialize;
use time::OffsetDateTime;
use utoipa::ToSchema;

use crate::domain::models::{
    AbsenceChild, AbsenceDayDefault, AbsenceEntry, AbsenceType, ActiveTimer, Activity, Project,
    TimeEntry, TimeEntryDayStatus, TimeEntryStatus, TimerHistoryEntry, WeeklyStats,
};

/// Response for the get timer endpoint.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GetTimerResponse {
    pub timer: Option<TimerResponse>,
}

/// Response for saving the active timer.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SaveTimerResponse {
    pub entry: TimeEntryResponse,
}

/// Active timer response - all timers are standalone now.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct TimerResponse {
    /// When the timer was started (ISO 8601).
    #[serde(with = "time::serde::rfc3339")]
    #[schema(value_type = String, format = "date-time")]
    pub start_time: OffsetDateTime,
    /// Project ID (if set).
    pub project_id: Option<String>,
    /// Project name (if set).
    pub project_name: Option<String>,
    /// Activity ID/code (if set).
    pub activity_id: Option<String>,
    /// Activity name (if set).
    pub activity_name: Option<String>,
    /// User note.
    pub note: String,
    /// Elapsed hours.
    pub hours: i64,
    /// Elapsed minutes (within current hour).
    pub minutes: i64,
    /// Elapsed seconds (within current minute).
    pub seconds: i64,
}

impl From<ActiveTimer> for TimerResponse {
    fn from(timer: ActiveTimer) -> Self {
        let (hours, minutes, seconds) = timer.elapsed_hms();
        Self {
            start_time: timer.started_at,
            project_id: timer.project_id.map(|id| id.to_string()),
            project_name: timer.project_name,
            activity_id: timer.activity_id.map(|id| id.to_string()),
            activity_name: timer.activity_name,
            note: timer.note,
            hours,
            minutes,
            seconds,
        }
    }
}

/// Project response - simplified for frontend use.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ProjectResponse {
    pub project_id: String,
    pub project_name: String,
}

impl From<Project> for ProjectResponse {
    fn from(project: Project) -> Self {
        Self {
            project_id: project.id.to_string(),
            project_name: project.name,
        }
    }
}

/// Activity response - simplified for frontend use.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ActivityResponse {
    /// Activity code (used in API calls).
    pub activity: String,
    /// Activity display name.
    pub activity_name: String,
}

impl From<Activity> for ActivityResponse {
    fn from(activity: Activity) -> Self {
        Self {
            activity: activity.id.to_string(),
            activity_name: activity.name,
        }
    }
}

/// Time entry response - completed time registration.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "lowercase")]
pub enum TimeEntryStatusResponse {
    Open,
    Approved,
    Certified,
}

impl From<TimeEntryStatus> for TimeEntryStatusResponse {
    fn from(status: TimeEntryStatus) -> Self {
        match status {
            TimeEntryStatus::Open => Self::Open,
            TimeEntryStatus::Approved => Self::Approved,
            TimeEntryStatus::Certified => Self::Certified,
        }
    }
}

#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct TimeEntryResponse {
    pub registration_id: String,
    pub project_id: String,
    pub project_name: String,
    pub activity_id: String,
    pub activity_name: String,
    /// Date in YYYY-MM-DD format.
    pub date: String,
    pub hours: f64,
    pub note: Option<String>,
    #[serde(with = "time::serde::rfc3339::option")]
    #[schema(value_type = Option<String>, format = "date-time")]
    pub start_time: Option<OffsetDateTime>,
    #[serde(with = "time::serde::rfc3339::option")]
    #[schema(value_type = Option<String>, format = "date-time")]
    pub end_time: Option<OffsetDateTime>,
    pub week_number: u8,
    /// Attestation status: `open`, `approved`, or `certified`.
    pub status: TimeEntryStatusResponse,
}

impl From<TimeEntry> for TimeEntryResponse {
    fn from(entry: TimeEntry) -> Self {
        Self {
            registration_id: entry.registration_id,
            project_id: entry.project_id.to_string(),
            project_name: entry.project_name,
            activity_id: entry.activity_id.to_string(),
            activity_name: entry.activity_name,
            date: entry.date.to_string(),
            hours: entry.hours,
            note: entry.note,
            start_time: entry.start_time,
            end_time: entry.end_time,
            week_number: entry.week_number,
            status: entry.status.into(),
        }
    }
}

/// Date-level time entry status response.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct TimeEntryDayStatusResponse {
    /// Date in YYYY-MM-DD format.
    pub date: String,
    /// Attestation status: `open`, `approved`, or `certified`.
    pub status: TimeEntryStatusResponse,
}

impl From<TimeEntryDayStatus> for TimeEntryDayStatusResponse {
    fn from(day_status: TimeEntryDayStatus) -> Self {
        Self {
            date: day_status.date.to_string(),
            status: day_status.status.into(),
        }
    }
}

/// Timer history entry response.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct TimerHistoryEntryResponse {
    pub id: i32,
    pub registration_id: Option<String>,
    pub user_id: i32,
    #[serde(with = "time::serde::rfc3339")]
    #[schema(value_type = String, format = "date-time")]
    pub start_time: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339::option")]
    #[schema(value_type = Option<String>, format = "date-time")]
    pub end_time: Option<OffsetDateTime>,
    pub project_id: Option<String>,
    pub project_name: Option<String>,
    pub activity_id: Option<String>,
    pub activity_name: Option<String>,
    pub note: Option<String>,
    #[serde(with = "time::serde::rfc3339")]
    #[schema(value_type = String, format = "date-time")]
    pub created_at: OffsetDateTime,
}

impl From<TimerHistoryEntry> for TimerHistoryEntryResponse {
    fn from(entry: TimerHistoryEntry) -> Self {
        Self {
            id: entry.id.as_i32(),
            registration_id: entry.registration_id,
            user_id: entry.user_id.as_i32(),
            start_time: entry.start_time,
            end_time: entry.end_time,
            project_id: entry.project_id.map(|p| p.to_string()),
            project_name: entry.project_name,
            activity_id: entry.activity_id.map(|a| a.to_string()),
            activity_name: entry.activity_name,
            note: entry.note,
            created_at: entry.created_at,
        }
    }
}

/// Weekly stats response.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct WeeklyStatsResponse {
    pub worked_hours: f64,
    pub scheduled_hours: f64,
    pub remaining_hours: f64,
    pub absence_hours: f64,
    pub covered_hours: f64,
    pub period_flex_hours: f64,
}

impl From<WeeklyStats> for WeeklyStatsResponse {
    fn from(info: WeeklyStats) -> Self {
        Self {
            worked_hours: info.worked_hours,
            scheduled_hours: info.scheduled_hours,
            remaining_hours: info.remaining_hours,
            absence_hours: info.absence_hours,
            covered_hours: info.covered_hours,
            period_flex_hours: info.period_flex_hours,
        }
    }
}

/// Absence entry response.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AbsenceEntryResponse {
    pub absence_id: String,
    pub date: String,
    pub hours: f64,
    pub absence_type: AbsenceType,
    pub absence_type_label: &'static str,
    pub child: Option<String>,
    pub comment: Option<String>,
    pub managed: bool,
    pub deletable: bool,
}

/// Available absence type response.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AbsenceTypeResponse {
    pub absence_type: AbsenceType,
    pub absence_type_label: &'static str,
}

/// Registered child available for child-related absence reporting.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AbsenceChildResponse {
    pub name: String,
    pub birth_date: Option<String>,
}

impl From<AbsenceType> for AbsenceTypeResponse {
    fn from(absence_type: AbsenceType) -> Self {
        Self {
            absence_type,
            absence_type_label: absence_type.label(),
        }
    }
}

impl From<AbsenceChild> for AbsenceChildResponse {
    fn from(child: AbsenceChild) -> Self {
        Self {
            name: child.name,
            birth_date: child.birth_date.map(|date| date.to_string()),
        }
    }
}

impl From<AbsenceEntry> for AbsenceEntryResponse {
    fn from(entry: AbsenceEntry) -> Self {
        Self {
            absence_id: entry.absence_id,
            date: entry.date.to_string(),
            hours: entry.hours,
            absence_type: entry.absence_type,
            absence_type_label: entry.absence_type.label(),
            child: entry.child,
            comment: entry.comment,
            managed: entry.managed,
            deletable: entry.deletable,
        }
    }
}

/// Default hours for one absence day.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AbsenceDayDefaultResponse {
    pub date: String,
    pub scheduled_hours: f64,
}

impl From<AbsenceDayDefault> for AbsenceDayDefaultResponse {
    fn from(day: AbsenceDayDefault) -> Self {
        Self {
            date: day.date.to_string(),
            scheduled_hours: day.scheduled_hours,
        }
    }
}
